Packs/Druva/Integrations/DruvaEventCollector/DruvaEventCollector.py

```diff
@@ -0,0 +1,289 @@
+            expiration_time = datetime.strptime(
+                cache["expiration_time"], DATE_FORMAT_FOR_TOKEN
+            )
+
+            # check if token is still valid, and use the old one. otherwise regenerate a new one
+            if (expiration_time - now).total_seconds() > 0:
+                demisto.debug("No need to regenerate the token, it is still valid")
+                self._set_headers(token)
+                return
+
+        demisto.debug("IntegrationContext token cache is empty or token has expired, regenerating a new token")
+        raw_token, expires_in_seconds = self._refresh_access_token()
+        self._set_headers(raw_token)
+
+        #  token["expires_in"] - 60 seconds for safety
+        set_integration_context(
+            {
+                "Token": raw_token,
+                "expiration_time": (
+                    now + timedelta(seconds=(expires_in_seconds - 60))
+                ).strftime(DATE_FORMAT_FOR_TOKEN),
+            }
+        )
+
+    def _refresh_access_token(self) -> tuple[str, int]:
+        """
+        Since the validity of the Access Token is 30 minutes, this method refreshes it and returns the new token json.
+        returns:
+            - the token
+            - the expiration in seconds
+        """
+        encoded_credentials = base64.b64encode(self.credentials.encode())
+        decoded_credentials = encoded_credentials.decode("utf-8")
+
+        headers = {
+            "Content-Type": "application/x-www-form-urlencoded",
+            "Authorization": f"Basic {decoded_credentials}",
+        }
+        data = {"grant_type": "client_credentials", "scope": "read"}
+
+        try:
+            response_json = self._http_request(
+                method="POST", url_suffix="/token", headers=headers, data=data
+            )
+        except Exception as e:
+            # 400 - "invalid_grant" - reason: invalid Server URL, Client ID or Secret Key.
+            if "invalid_grant" in str(e):
+                raise DemistoException(
+                    "Error in test-module: Make sure Server URL, Client ID and Secret Key are correctly entered."
+                ) from e
+            raise
+        return response_json["access_token"], response_json["expires_in"]
+
+    def search_events(self, tracker: Optional[str] = None) -> dict:
+        """
+        Searches for Druva events.
+
+        Args:
+            tracker: pointer to the last event we got last time
+
+        Returns:
+            dict: List of events
+        """
+        demisto.debug(f'DEC: This is the tracker before encoding: {tracker=}')
+
+        if tracker:
+            encoding_tracker = quote(tracker, safe="!~*'()")  # remove invalid characters from tracker
+            demisto.debug(f'DEC: This is the tracker after encoding: {encoding_tracker=}')
+        else:
+            encoding_tracker = ""
+
+        url_suffix_tracker = f"?tracker={encoding_tracker}" if encoding_tracker else ""
```

Comment:
instead of this `if`, we can put `url_suffix_tracker =` in the `if` above, and `url_suffix_tracker = ""` in the `else`

Reply:
Done